specification/dsdl/serializable_types.tex

```diff
@@ -286,3 +286,41 @@ \subsubsection{Unions}\label{sec:dsdl_composite_tagged_unions}
 
 The value of a tagged union object is a function of the field attribute which value it is currently holding
 and the value of the field attribute itself.
+
+\subsubsection{Type polymorphism}
+
+Type polymorphism is supported in DSDL through structural subtyping.
+The following definition relies on the concept of \emph{field attribute},
+which is introduced in section \ref{sec:dsdl_attributes}.
+
+Let $B$ and $D$ be DSDL message types whose data schemas define $b$ and $d$ field attributes, respectively.
+Let each field attribute be assigned a sequential index according to its position in the DSDL definition
+(see section \ref{sec:dsdl_grammar} on statement ordering).
+
+$D$ is a \emph{structural subtype} of $B$ if, and only if,
+$b \leq d$, $B$ is not $D$, neither $B$ nor $D$ define a tagged union, and
+for each field attribute of $B$ at index $i$ there is an equal field attribute in $D$ at index $i$.
+
+If $D$ is a structural subtype of $B$, then $B$ is a structural supertype of $D$.
+
+\begin{remark}
+    Subtyping example. First type:
+
+    \begin{minted}{python}
+        float64 a       # Index 0
+        int16[<=9] b    # Index 1
+    \end{minted}
+
+    The second type is a structural subtype of the first type:
+
+    \begin{minted}{python}
+        float64 a       # Index 0
+        int16[<=9] b    # Index 1
+        uint8 foo       # Index 2
+    \end{minted}
+
+    A message type that defines zero fields is a structural supertype of any other message type.
+    A message type may have an arbitrary number of supertypes as long as the field equality constraint is satisfied.
+
+    Polymorphic relations are not defined on service types and tagged unions.
```

Comment:
A type hierarchy is an ordered set of types such that for each pair of its members one type is a subtype of another,
and for any member its supertypes are located on the left.

Given a publisher P and a subscriber S, where P emits objects of type X, and S expects objects of type Y, we say that their publish-subscribe relation is:

- Type-covariant if X is a supertype of Y and X is not Y.
- Type-contravariant if X is a subtype of Y and X is not Y.
- Type-invariant if X is Y.
- Type-orthogonal if Y and X belong to different type hierarchies.

For example, a subscriber expecting `uavcan.primitive.Empty.1.0` is type-covariant with any publisher.

Due to the bidirectional nature of service exchanges, they can be either type-invariant or type-orthogonal, which renders structural subtyping much less useful. To work around that, we would have to reason about service requests and service responses as independent members of a type hierarchy. I am envisioning simple aliases for request and response schemas that we could add later, like `uavcan.file.Read.Request.1.0` (notice `Request`), each being an independent type. For now, I suggest to stash this into the TODO list and move on.

> Similarly, are you implying a prohibition on polymorphic types within tagged unions or just on the unions themselves?

I don't understand the difference between tagged unions and just unions. They are the same thing in my mind.

Regardless, this question forced me to think about this a bit more. The current prohibition derives from the common practice employed in major programming languages to not define polymorphic relations on unions (C++, D, Rust (enum or union)). UAVCAN, however, is different because it does not have any built-in means of dynamic polymorphism, and unions can be used to work around that (like variant types). It is useful to define subtyping such that a union is a subtype of another iff the type of each field of the subtype is of the same type or a subtype of the corresponding field in the supertype, the subtype contains all fields of the supertype at the same indexes, and the bit width of the union tag field is the same across the hiearchy. This is clunky but I will work out a cleaner definition for the spec. For example:

Base:

```python
@union
uavcan.primitive.Empty.1.0 foo
float16 bar
uint8 zoo
```

Subtype:

```python
@union
uavcan.diagnostic.Record.1.0 foo    # Subtype
float16 bar                         # Same
uint8 zoo                           # Same
int64[<=64] baz                     # New field
```

The general idea is to retain the full wire compatibility so that a subtype can be used to decode any instance of any supertype.

I am going to add this in my next commit and then we can review it.

---

I want to share a general observation I made while reviewing your feedback. The data type compatibility concepts as we have defined them no longer work with the new methods of ensuring compatible type extension -- namely, polymorphism, implicit truncation, and implicit zero extension. Same applies to versioning: addition of an optional field shall be allowed if its correct usage can be enabled by the implicit truncation/extension rules.

We shall discard the current data type compatibility policies and define a more compact set from scratch. I want to do that in a separate pull request, however.

Reply:
See https://github.com/UAVCAN/specification/pull/64/commits/28902afd2aefa1e9e7d5b93aa1d642940b2bbdf7